src/coroutine.rs

```diff
@@ -0,0 +1,218 @@
+        self.future.take();
+    }
+
+    fn __await__(self_: Py<Self>) -> Py<Self> {
+        self_
+    }
+
+    fn __next__(&mut self, py: Python<'_>) -> PyResult<IterNextOutput<PyObject, PyObject>> {
+        self.poll(py, None)
+    }
+}
+
+/// `asyncio.Future` wrapper, implementing [`ArcWake`] by calling `Future.set_result`.
+///
+/// It embeds the event loop thread id to determine if it's necessary to use `call_soon_threadsafe`.
+struct PyFutureWaker {
+    future: PyObject,
+    thread_id: ThreadId,
+}
+
+impl PyFutureWaker {
+    fn wake_same_thread(&self, py: Python<'_>) {
+        // `set_result` can be called directly because in the event loop thread
+        self.future
+            .call_method1(py, intern!(py, "set_result"), (py.None(),))
+            .expect("error while calling Future.set_result");
+    }
+
+    fn wake_threadsafe(&self, py: Python<'_>) {
+        // `set_result` must be called in event loop thread, so it requires `call_soon_threadsafe`
+        let set_result = self
+            .future
+            .getattr(py, intern!(py, "set_result"))
+            .expect("error while calling Future.set_result");
+        let event_loop = self
+            .future
+            .call_method0(py, intern!(py, "get_loop"))
+            .expect("error while calling Future.get_loop");
+        if let err @ Err(_) = event_loop.call_method1(
+            py,
+            intern!(py, "call_soon_threadsafe"),
+            (set_result, py.None()),
+        ) {
+            // `call_soon_threadsafe` will raise if the event loop is closed;
+            // instead of catching an unspecific `RuntimeError`, check directly if it's closed.
+            let is_closed = event_loop
+                .call_method0(py, "is_closed")
+                .and_then(|res| res.extract(py))
+                .unwrap_or(false);
+            // if the loop is not closed, panic
+            if !is_closed {
+                err.expect("error while calling EventLoop call_soon_threadsafe");
+            }
+        }
+    }
+}
+
+impl ArcWake for PyFutureWaker {
+    fn wake_by_ref(arc_self: &Arc<Self>) {
+        // Choose appropriate wake method depending on wether it is called on the same thread
+        // than the event loop
+        if current_thread_id() == arc_self.thread_id {
+            Python::with_gil(|gil| arc_self.wake_same_thread(gil))
+        } else {
+            Python::with_gil(|gil| arc_self.wake_threadsafe(gil))
+        }
+    }
+}
+
+/// Instantiate an `asyncio.Future` in the current event loop.
+fn asyncio_future(py: Python<'_>) -> PyResult<PyObject> {
+    static FUTURE: GILOnceCell<PyObject> = GILOnceCell::new();
+    let import_future = || -> PyResult<_> {
+        let module = py.import("asyncio")?;
+        Ok(module.getattr("Future")?.into())
+    };
+    FUTURE.get_or_try_init(py, import_future)?.call0(py)
+}
+
+/// Alternative to `std::thread::current().id()` without unnecessary Arc clone + drop.
+pub(crate) type ThreadId = usize;
```

Comment:
I am very much not a fan of implementing our own thread ID. Even if we do decided that this is a worthwhile performance optimization, I think it should be done in a separate change to keep this as simple as possible.

Reply:
Or alternatively, I don't think we we need thread ID at all. We just want to know if the current thread hosts the event loop, not which one actually is, i.e. a thread-local flag should suffice, e.g.

```rust
thread_local! {
  static HOSTS_EVENT_LOOP: Cell<bool> = const { Cell::new(false) };
}
```

which we set to `true` when the waker is constructed.

(Note that this should also conditionally use `const` syntax as we already do in

https://github.com/PyO3/pyo3/blob/2b91fea94f4f6371b995f4c649c1ea80bd5f4dd2/src/gil.rs#L15

but this might also better be a optimization follow up.)

Reply:
As written in reply to your other comment, this thread id "optimization" is wrong and will be removed.